### PR TITLE
Sequential unwrapPhase stitcher

### PR DESCRIPTION
**Pull request by mgovorcin**
adding new method for stitching frames of unwrapped phase along the track.

Sequential stitching is perform with forward and backward 2-pi integer cycle and range corrections using overlapping connected components between two neighboring frames

### CODE REVIEW

File: tools/ARIAtools/extractProduct.py [hunk 1]
```diff
@@ -87,7 +87,7 @@ def createParser():
         help='Amplitude threshold below which to mask. Specify "None" to not use amplitude mask. By default "None".')
     parser.add_argument('-nt', '--num_threads', dest='num_threads', default='2', type=str,
         help='Specify number of threads for multiprocessing operation in gdal. By default "2". Can also specify "All" to use all available threads.')
-#    parser.add_argument('-sm', '--stitchMethod', dest='stitchMethodType',  type=str, default='overlap', help="Method applied to stitch the unwrapped data. Either 'overlap', where product overlap is minimized, or '2stage', where minimization is done on connected components, are allowed methods. Default is 'overlap'.")
+    parser.add_argument('-sm', '--stitchMethod', dest='stitchMethodType',  type=str, default='overlap', help="Method applied to stitch the unwrapped data. Allowed methods are: 'overlap', '2stage', and 'sequential'. Default is 'overlap'.")
```
Comment (dbekaert) [inline tools/ARIAtools/extractProduct.py:90]: You might want to keep some sort of description in here about the method.

File: tools/ARIAtools/sequential_stitching.py [hunk 2]
```diff
@@ -0,0 +1,945 @@
...
+# Copyright 2023, by the California Institute of Technology. ALL RIGHTS
+# RESERVED. United States Government Sponsorship acknowledged.
+#
+#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+import warnings
+import numpy as np
+from numpy.typing import NDArray, ArrayLike
+
+from typing import List, Optional, Tuple, Union
```
Comment (dbekaert) [inline tools/ARIAtools/sequential_stitching.py:14]: just double checking if these are default requirments upon install.

File: tools/ARIAtools/sequential_stitching.py [hunk 3]
```diff
@@ -0,0 +1,945 @@
...
+    -------
+    diff_value : float
+        mean offset between overlapping comoponent in two frames
+    correction2pi : float
+        2pi interger cycles between overlapping comoponent in two frames
+    range_corr : float
+        correction for non 2-pi shift between overlapping comoponent in two frames 
+    """  
+    # Not sure if copy() is needed, left it here to avoid overwriting array due to numpy referencing
+    # TODO: use np.where to find the component in the data and keep overlap dimensions for comparison
```
Comment (dbekaert) [inline tools/ARIAtools/sequential_stitching.py:275]: is this completed or still placeholder?

**Review by dbekaert — APPROVED** [review]
LGTM just minor comments.

**Comment by dbekaert** [comment]
@sssangha @cmarshak Can you advice on the github action tests. Some seems to be cancelled others to be failed. These seems to be related to credentials? 
```Username/Password Authentication Failed.```
``` Error: The operation was canceled.```

Anything specific that needs to be set-up such other users also have the test run to completion?

File: tools/ARIAtools/sequential_stitching.py (hunk 2 above)
Comment (cmarshak) [reply tools/ARIAtools/sequential_stitching.py:14]: `typing` is in the standard library. Helpful for writing `dataclasses` and self-documenting functions e.g. `get_month(input_date: Union[datetime.datetime, str])`.

**Comment by cmarshak** [comment]
> @sssangha @cmarshak Can you advice on the github action tests. Some seems to be cancelled others to be failed. These seems to be related to credentials? `Username/Password Authentication Failed.` ` Error: The operation was canceled.`
> 
> Anything specific that needs to be set-up such other users also have the test run to completion?

So, Marin is forking this repo and testing it on his own branch. The netrc that's generated for tests requires the repository secrets. I am not sure how that's shared on forks. If he does PR using branches within the repo that will work. I can google if there are solutions. Likely add the relevant secrets in his forked repo. Here are the naming conventions: https://github.com/mgovorcin/ARIA-tools/blob/dev/.github/workflows/test.yml#L21-L25

As in you would use `EARTHDATA_USERNAME` and put in your earthdata username.

**Comment by cmarshak** [comment]
image

`Anyone with collaborator access to the repositories with access to a secret or variable can use it for Actions. They are not passed to workflows that are triggered by a pull request from a fork.`

So either:

a) make PRs (from collaborators/developors) directly within ARIA-tools branches
b) put in github secrets to forked repository as above

I honestly would recommend doing a) for development work because it will be faster (particularly if and when you decide to add more tests and/or static analysis to github actions). For b), just make sure you keep your fork after you configure the secrets so you can continue adding. You might want to make some instructions in the readme for such PRs.

**Comment by cmarshak** [comment]
Lastly, "cancelled actions" occur if one or more of the same tests/actions in a matrix fail e.g. if python 3.8 - 3.10 are tested and 3.9 fails (as above), then 3.8 and 3.10 tests will get cancelled.

Here, matrix refers to this: https://github.com/aria-tools/ARIA-tools/blob/dev/.github/workflows/test.yml#L14-L16

File: tools/ARIAtools/sequential_stitching.py (hunk 3 above)
Comment (mgovorcin) [reply tools/ARIAtools/sequential_stitching.py:275]: It is completed, the note is for future clean-ups

**Comment by mgovorcin** [comment]
Closing this PR and moving to PR #351 to avoid test fails due to not passing credentials from fork

**Pull request closed without merging**